pymacaroons/serializers/binary_serializer.py

```diff
@@ -113,3 +236,103 @@ def _depacketize(self, packet):
         key = packet.split(b' ')[0]
         value = packet[len(key) + 1:-1]
         return (key, value)
+
+
+def _append_packet(data, field_type, packet_data=None):
+    _encode_uvarint(data, field_type)
+    if field_type != _EOS:
+        _encode_uvarint(data, len(packet_data))
+        data.extend(packet_data)
+
+
+def _parse_section_v2(data):
+    ''' Parses a sequence of packets in data.
+
+    The sequence is terminated by a packet with a field type of EOS
+    :param data bytes to be deserialized.
+    :return: the rest of data and an array of packet V2
+    '''
+
+    from pymacaroons.exceptions import MacaroonDeserializationException
+
+    prev_field_type = -1
+    packets = []
+    while True:
+        if len(data) == 0:
+            raise MacaroonDeserializationException(
+                'section extends past end of buffer')
+        rest, packet = _parse_packet_v2(data)
+        if packet.field_type == _EOS:
+            return rest, packets
+        if packet.field_type <= prev_field_type:
+            raise MacaroonDeserializationException('fields out of order')
+        packets.append(packet)
+        prev_field_type = packet.field_type
+        data = rest
+
+
+def _parse_packet_v2(data):
+    ''' Parses a V2 data packet at the start of the given data.
+
+    The format of a packet is as follows:
+
+    field_type(varint) payload_len(varint) data[payload_len bytes]
+
+    apart from EOS which has no payload_en or data (it's a single zero byte).
+
+    :param data:
+    :return: rest of data, PacketV2
+    '''
+    from pymacaroons.exceptions import MacaroonDeserializationException
+
+    ft, n = _decode_uvarint(data)
+    data = data[n:]
+    if ft == _EOS:
+        return data, PacketV2(ft, None)
+    payload_len, n = _decode_uvarint(data)
+    data = data[n:]
+    if payload_len > len(data):
+        raise MacaroonDeserializationException(
+            'field data extends past end of buffer')
+    return data[payload_len:], PacketV2(ft, data[0:payload_len])
+
+
+def _encode_uvarint(data, n):
+    ''' Encodes integer into variable-length format into data.'''
+    if n < 0:
+        raise ValueError('only support positive integer')
+    while True:
+        this_byte = n & 127
+        n >>= 7
+        if n == 0:
+            data.append(this_byte)
+            break
+        data.append(this_byte | 128)
+
+
+def _decode_uvarint(data):
+    ''' Decode a variable -length integer.
+
+    Reads a sequence of unsigned integer byte and decodes them into an integer
+    in variable-length format and returns it and the length read.
+    '''
+    n = 0
+    shift = 0
+    length = 0
+    for b in data:
+        if not isinstance(b, int):
+            b = six.byte2int(b)
+        if b < 0x80:
+            return n | b << shift, length + 1
+        n |= (b & 0x7f) << shift
+        shift += 7
+        length += 1
+    return n, length
+
+
+def _is_ascii_hex(b):
+    if b >= b'0' and b <= b'9':
```

Comment:
This doesn't seem quite right to me - it's comparing an integer with a string.
I think you might mean:

     if b >= ord('0') and b <= ord('9'):

and similarly below.

I'd suggest adding a test that forces the code to go through these paths,
because I'm pretty sure this would return the wrong result in some circumstances.